src/Rules/Functions/ParameterCastableToStringFunctionRule.php

```diff
@@ -0,0 +1,170 @@
+use PHPStan\Reflection\ReflectionProvider;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPStan\Rules\RuleLevelHelper;
+use PHPStan\Type\ErrorType;
+use PHPStan\Type\Type;
+use PHPStan\Type\VerbosityLevel;
+use function array_key_exists;
+use function count;
+use function in_array;
+use function sprintf;
+
+/**
+ * @implements Rule<Node\Expr\FuncCall>
+ */
+class ParameterCastableToStringFunctionRule implements Rule
+{
+
+	public function __construct(
+		private ReflectionProvider $reflectionProvider,
+		private RuleLevelHelper $ruleLevelHelper,
+	)
+	{
+	}
+
+	public function getNodeType(): string
+	{
+		return FuncCall::class;
+	}
+
+	public function processNode(Node $node, Scope $scope): array
+	{
+		if (!($node->name instanceof Node\Name)) {
+			return [];
+		}
+
+		if (!$this->reflectionProvider->hasFunction($node->name, $scope)) {
+			return [];
+		}
+
+		$functionReflection = $this->reflectionProvider->getFunction($node->name, $scope);
+		$functionName = $functionReflection->getName();
+		$implodeFunctions = ['implode', 'join'];
+		$checkAllArgsFunctions = ['array_intersect', 'array_intersect_assoc', 'array_diff', 'array_diff_assoc'];
+		$checkFirstArgFunctions = [
+			'array_unique',
+			'array_combine',
+			'sort',
+			'rsort',
+			'asort',
+			'arsort',
+			'natcasesort',
+			'natsort',
+			'array_count_values',
+			'array_fill_keys',
+		];
+
+		if (
+			!in_array($functionName, $checkAllArgsFunctions, true)
+			&& !in_array($functionName, $checkFirstArgFunctions, true)
+			&& !in_array($functionName, $implodeFunctions, true)
+		) {
+			return [];
+		}
+
+		$origArgs = $node->getArgs();
+		$parametersAcceptor = ParametersAcceptorSelector::selectFromArgs(
+			$scope,
+			$origArgs,
+			$functionReflection->getVariants(),
+			$functionReflection->getNamedArgumentsVariants(),
+		);
+
+		$errorMessage = 'Parameter %s of function %s expects an array of values castable to string, %s given.';
+		$getNormalizedArgs = static function () use ($parametersAcceptor, $node): array {
+			$normalizedFuncCall = ArgumentsNormalizer::reorderFuncArguments($parametersAcceptor, $node);
+
+			if ($normalizedFuncCall === null) {
+				return [];
+			}
+
+			return $normalizedFuncCall->getArgs();
+		};
+		if (in_array($functionName, $implodeFunctions, true)) {
+			$normalizedArgs = $getNormalizedArgs();
+			$errorMessage = 'Parameter %s of function %s expects array<string>, %s given.';
+			if (count($normalizedArgs) === 1) {
+				$argsToCheck = [0 => $normalizedArgs[0]];
+			} elseif (count($normalizedArgs) === 2) {
+				$argsToCheck = [1 => $normalizedArgs[1]];
+			} else {
+				return [];
+			}
+		} elseif (in_array($functionName, $checkAllArgsFunctions, true)) {
+			$argsToCheck = $origArgs;
+		} elseif (in_array($functionName, $checkFirstArgFunctions, true)) {
+			$normalizedArgs = $getNormalizedArgs();
+			if ($normalizedArgs === []) {
+				return [];
+			}
+			$argsToCheck = [0 => $normalizedArgs[0]];
```

Comment:
What if the index does not exist?

Reply:
I didn't think of that. But it seems that this cannot happen on non-empty `$normalizedArgs`, because the normalizer only fills optional parameter holes, if a required parameter is missing it seems to return null. I guess it doesn't hurt to be extra defensive and check it anyway.